### DISCUSSION TITLE
[feature] Devcontainers support

### DISCUSSION THREAD

**Discussion by IgnisDa** [helix-editor/helix#10248] [General]
Hello. Thank you developers and awesome project!

It would be really great if helix support developing in devcontainers (either in core or via plugins). A lot of developers prefer not to clutter their systems with language tools (like rust-analyzer), and devcontainers does really well in addressing.

Looking at it from a surface level, it should be doable. Devcontainers has a [CLI](https://github.com/devcontainers/cli) that can be packaged and installed once [this](https://github.com/devcontainers/cli/pull/343) PR lands.

Helix itself can be run similar to how their [vim example](https://github.com/devcontainers/cli/tree/main/example-usage/tool-vim-via-ssh) works.

Would be great if you could tell me the viability of this feature. Thanks!

**Comment by kirawi** [comment]
Duplicate of https://github.com/helix-editor/helix/issues/3002

**Comment by IgnisDa** [comment]
@kirawi The above issue runs the LSP inside a container. The feature I want support for runs Helix in the container too. Thus I do not think this is a duplicate.

**Comment by IgnisDa** [comment]
After some hacking, here is the progress I made:

- The base image is taken from [this](https://github.com/IgnisDa/developrs/blob/589f510a9ddb44c0adc0f1fb1a4d1d1c2fc60f0b/docker/Dockerfile) branch
- The changes required to my existing devcontainer are in [this](https://github.com/IgnisDa/rust-libs/tree/sshd) branch

To get it running, clone the repo in step 1, checkout the correct branch and then run the command in `docker/README.md`. Then clone repo in step 2, checkout correct branch and run `./.helix/start.sh`. ~It will prompt you for a password, type `password` there~ and helix should run. I have already installed `rust-analyzer` in the image so the project should open up with a working LSP.

Things that need to be figured out here:
- [x] Remove password prompt. The scripts are taken from [here](https://github.com/devcontainers/cli/tree/main/example-usage/tool-vim-via-ssh) but the password prompt still comes up. (fixed in https://github.com/IgnisDa/rust-libs/commit/043bb64ca79b3e0d166930c2082c56a0d6772793)
- [ ] Figure out how to forward github credentials. Forwarding SSH credentials seem to be straightforward (tutorial [here](https://docs.github.com/en/developers/overview/using-ssh-agent-forwarding)). But I use `gh auth login` and would like the same credentials to work.

**Comment by IgnisDa** [comment]
I made a pubic gist with info on how to get basic devcontainer support working: https://gist.github.com/IgnisDa/f4c1ffb3b798686df3193a80cbd5b3eb.

**Comment by belowm** [comment]
Since helix is very lightweight and doesn't require an X server, what about running helix itself in a container together with the lsp and tooling for the language in question?

**Comment by YeungOnion** [comment]
One thing I'd like to note is that if helix is running on the container, if that container is remote then you may have notable latency. At some point, I looked into using [distant](https://distant.dev/about/overview/) but I decided to put my efforts elsewhere and just use VSCode for devcontainer development since it got the job done. Esp with neovim integrations.
